lib/intrinsics/src/printf.jl

```diff
@@ -38,18 +38,84 @@ end
         llvm_f, _ = create_function(T_int32, param_types)
         mod = LLVM.parent(llvm_f)
 
-        # generate IR
         IRBuilder() do builder
             entry = BasicBlock(llvm_f, "entry")
             position!(builder, entry)
 
+            T_actual_args = LLVMType[]
+            actual_args = LLVM.Value[]
+            for (i, (arg, argtyp, argval)) in enumerate(zip(parameters(llvm_f), arg_types, argspec))
+                # if the value is a Val, we'll try to emit it as a constant
+                const_arg = if argval <: Val
+                    # also pass the actual value for the fallback path (and to simplify
+                    # construction of the LLVM function, where we can ignore constants)
+                    argexprs[i] = argval.parameters[1]
+
+                    argval.parameters[1]
+                else
+                    nothing
+                end
```

Comment:
I don't think we need all of the features of `@typed_ccall`, only the `Ptr`-related ones.